src/installer/corehost/corehost.cpp

```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
 
     // Obtain the entrypoints.
     int rc;
-    hostfxr_main_startupinfo_fn main_fn_v2 = reinterpret_cast<hostfxr_main_startupinfo_fn>(pal::get_symbol(fxr, "hostfxr_main_startupinfo"));
-    if (main_fn_v2 != nullptr)
+#if defined(FEATURE_APPHOST)
+    if (bundle_marker_t::is_bundle())
     {
+        hostfxr_main_bundle_startup_info_fn main_fn_v3 = reinterpret_cast<hostfxr_main_bundle_startup_info_fn>(pal::get_symbol(fxr, "hostfxr_main_bundle_startup_info"));
+        if (main_fn_v3 == nullptr)
+        {
+            trace::error(_X("The required library %s does not support single-file apps."), fxr_path.c_str());
+            rc = StatusCode::CoreHostEntryPointFailure;
+        }
+
         const pal::char_t* host_path_cstr = host_path.c_str();
         const pal::char_t* dotnet_root_cstr = dotnet_root.empty() ? nullptr : dotnet_root.c_str();
         const pal::char_t* app_path_cstr = app_path.empty() ? nullptr : app_path.c_str();
+        int64_t bundle_header_offset = bundle_marker_t::header_offset();
 
-        trace::info(_X("Invoking fx resolver [%s] v2"), fxr_path.c_str());
+        trace::info(_X("Invoking fx resolver [%s] v3"), fxr_path.c_str());
         trace::info(_X("Host path: [%s]"), host_path.c_str());
         trace::info(_X("Dotnet path: [%s]"), dotnet_root.c_str());
         trace::info(_X("App path: [%s]"), app_path.c_str());
+        trace::info(_X("Bundle Header Offset: [%lx]"), bundle_header_offset);
 
         hostfxr_set_error_writer_fn set_error_writer_fn = reinterpret_cast<hostfxr_set_error_writer_fn>(pal::get_symbol(fxr, "hostfxr_set_error_writer"));
         {
             propagate_error_writer_t propagate_error_writer_to_hostfxr(set_error_writer_fn);
 
-            rc = main_fn_v2(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr);
+            rc = main_fn_v3(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr,bundle_header_offset);
 
             if (trace::get_error_writer() != nullptr && rc == static_cast<int>(StatusCode::FrameworkMissingFailure) && !set_error_writer_fn)
```

Comment:
This is not necessary due to the above "version" check. This was introduced here: https://github.com/dotnet/runtime/pull/1222
The reason was to get a good error message for UI apps when running 3.0 apps on 2.0 only hostfxr which doesn't support error writer redirection. In this case the failure will happen earlier and if we get here it will be with >=5.0 hostfxr which will support the fully featured error reporting.

Maybe add comments describing this reasoning here and to the case below for 3.0 app on 2.0 hostfxr.